hono-client-kafka/src/main/java/org/eclipse/hono/client/kafka/impl/KafkaBasedDownstreamSender.java

```diff
@@ -0,0 +1,187 @@
+     * Constructor.
+     *
+     * @param closeHandler The handler to be called when {@link #close(Handler)} is invoked.
+     */
+    protected KafkaBasedDownstreamSender(final Handler<Void> closeHandler) {
+        this.closeHandler = closeHandler;
+    }
+
+    private List<KafkaHeader> getHeadersForProducerRecord(final String deviceId, final QoS qos,
+            final String contentType, final Map<String, ?> properties) {
+
+        final HashMap<String, Object> headerProperties = new HashMap<>();
+        if (properties != null) {
+            headerProperties.putAll(properties);
+        }
+
+        headerProperties.put(MessageHelper.SYS_PROPERTY_CONTENT_TYPE, contentType);
+        headerProperties.put(MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
+        headerProperties.put(MessageHelper.APP_PROPERTY_QOS, qos.ordinal());
+
+        if (headerProperties.containsKey(MessageHelper.APP_PROPERTY_DEVICE_TTD)
+                && !headerProperties.containsKey(MessageHelper.SYS_PROPERTY_CREATION_TIME)) {
+            // TODO set this as creation time in the KafkaRecord?
+
+            // must match http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-timestamp
+            // as defined in https://www.eclipse.org/hono/docs/api/telemetry/#forward-telemetry-data
+            final long timestamp = Instant.now().toEpochMilli();
+            headerProperties.put(MessageHelper.SYS_PROPERTY_CREATION_TIME, timestamp);
+        }
+
+        return getKafkaHeaders(headerProperties);
+    }
+
+    /**
+     * Sends a message downstream with the given producer.
+     *
+     * @param producer The Kafka producer to send the message.
+     * @param topic The topic to send the mesage to.
+     * @param deviceId The ID of the device that the data originates from.
+     * @param qos The delivery semantics to use for sending the data.
+     * @param contentType The content type of the data.
+     * @param payload The data to send.
+     * @param properties Additional meta data that should be included in the downstream message.
+     * @param context The currently active OpenTracing span (may be {@code null}). An implementation should use this as
+     *            the parent for any span it creates for tracing the execution of this operation.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future will be succeeded if the message has been sent downstream.
+     *         <p>
+     *         The future will be failed with a {@link org.eclipse.hono.client.ServerErrorException} if the data could
+     *         not be sent. The error code contained in the exception indicates the cause of the failure.
+     */
+    protected Future<Void> send(final KafkaProducer<String, Buffer> producer, final String topic,
+            final String deviceId, final QoS qos, final String contentType, final Buffer payload,
+            final Map<String, ?> properties, final SpanContext context) {
```

Comment:
I noticed that the `SpanContext` is not used yet. Regarding tracing I think we should take a look at https://github.com/opentracing-contrib/java-kafka-client. By using e.g. the interceptor based approach it seems like a quick win to enable distributed tracing for Kafka.

Reply:
I added tracing. I did not use the library you mentioned, because does not add much to the tracing logic we already have in place and does not provide the flexibility to adapt to our needs